### PR TITLE
Spec: Add a new virtual-customizable `home` attribute

### PR DESCRIPTION
**Pull request by chuckatkins**
Details and use cases covered in the new `Package attributes` section of the `Packaging Guide`

### CODE REVIEW

**Comment by chuckatkins** [comment]
Converting back to draft while I add some tests and resolve the issues they reveal

**Comment by chuckatkins** [comment]
@becker33 this is good to go now

File: lib/spack/docs/packaging_guide.rst
Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: Not sure how to convey that this could be `spec.prefix.lib` or `spec.prefix.lib64`, or something else recursively

File: lib/spack/docs/packaging_guide.rst
Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: Technically these are spec attributes (`self.spec.foo`), not package attributes (`self.spec.package.foo`)

File: lib/spack/docs/packaging_guide.rst
Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: Can't believe we never documented any of these before lol

**Review by adamjstewart — COMMENTED** [review]
I like it. It follows what I've already done with `libs`/`headers`/`command`. I've been using `home` in all of the Java providers and in Python, but we can rename that to `root`. The question is how do we convince people to start using `spec.root` instead of `prefix`? I still have to explain to people why they can't use `prefix.lib` because it might be in `prefix.lib64`. Telling people not to use `prefix` at all will be a tougher sell.

File: lib/spack/docs/packaging_guide.rst
Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst]: I suppose it's neither since their attributes on the `SpecBuildInterface`.  I could just change the section header to "Custom attributes", would that be sufficient?

File: lib/spack/docs/packaging_guide.rst
Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst]: I'll add some wording accordingly.

**Comment by chuckatkins** [comment]
> The question is how do we convince people to start using spec.root instead of prefix?  I still have to explain to people why they can't use prefix.lib because it might be in prefix.lib64. Telling people not to use prefix at all will be a tougher sell.

I don't think it needs to be a concentrated campaign or anything.  Updating examples in the documentation along with incorporating it into code reviews should suffice I think.  Packages can gradually make the change rather than one big sweep.

**Comment by adamjstewart** [comment]
This PR may also be useful for `IntelPackage` where the actual `bin`/`lib`/`include` dirs are a subdir of `prefix`.

File: lib/spack/spack/test/install.py
Comment (becker33) [inline lib/spack/spack/test/install.py]: Is `.normalized()` necessary here? I can't think what it would be achieving.

File: lib/spack/spack/test/install.py
Comment (becker33) [inline lib/spack/spack/test/install.py]: This is unnecessary, it's handled by the `install_mockery` fixture

File: var/spack/repos/builtin.mock/packages/attributes-foo/package.py
Comment (becker33) [inline var/spack/repos/builtin.mock/packages/attributes-foo/package.py]: No reason to change it, but just a note you can also use `self.spec.prefix.baz`

**Review by becker33 — DISMISSED** [review]
Two small changes, but generally this LGTM and I think it is the correct model for how to handle it in Spack.

File: lib/spack/docs/packaging_guide.rst
Comment (tgamblin) [inline lib/spack/docs/packaging_guide.rst]:
```suggestion
Often, a package will need to provide attributes for dependents to query
details about what it provides.  While any number of custom defined
```

File: lib/spack/docs/packaging_guide.rst
Comment (tgamblin) [inline lib/spack/docs/packaging_guide.rst]:
```suggestion
attributes can be implemented by a package, the four specific attributes
```

File: lib/spack/docs/packaging_guide.rst
Comment (tgamblin) [inline lib/spack/docs/packaging_guide.rst]: why not write the `headers` and `libs` defaults in terms of `spec.root`, so that if you change the root for a virtual, you automatically get appropriately implemented `headers` and `libs` attributes too?  I think this behavior will be more common, and you could still override both `foo_root` and `foo_headers` if you need to, but more likely you'll only need to override `foo_root`.

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/install.py
Comment (tgamblin) [reply lib/spack/spack/test/install.py]: Also we'll likely get rid of `normalized()` entirely some day soon (it's only really a concept within the old concretizer).

Comment (chuckatkins) [reply lib/spack/spack/test/install.py]: `.normalized()` was necessary to get to the underlying package spec  since the only direct dependencies are the virtual packages.  For the sake of the test though it doesn't really matter and I can setup the dependencies a bit different and get rid of `.normalize`

File: lib/spack/docs/packaging_guide.rst
Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst]: Why a single command, rather than a list? Also, need to check the code blow, but is `spec.command` supposed to be a Boolean?

File: lib/spack/docs/packaging_guide.rst
Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst]: Can you explain this, or point to relevant discussion that I might have missed? I think everything we have so far is implemented in terms of `spec.prefix`, so if `spec.root` is preferred this might be a maintenance issue.

File: lib/spack/docs/packaging_guide.rst
Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst:2858]: Can we make the example be based on a real use case (which I am sure exists)? By just reading the docs I don't understand which use case this PR covers (given that `mpi_libs` etc. already exist as properties).

File: lib/spack/spack/test/install.py
Comment (alalazo) [inline lib/spack/spack/test/install.py:56]: Would it be easy to break this single test into multiple parametrized tests, and make the body shorter when doing so? A unit test of 70 lines is harder to follow than a shorter one, and it's easier to miss bugs in the test itself.

File: var/spack/repos/builtin.mock/packages/attributes-foo/package.py
Comment (alalazo) [inline var/spack/repos/builtin.mock/packages/attributes-foo/package.py]: Why are we choosing this unusual layout? Is there a real reason behind it?

**Review by alalazo — CHANGES_REQUESTED** [review]
I left a few comments, let me know what you think about them.

File: lib/spack/docs/packaging_guide.rst
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: Ok, I see this is documenting the already existing `command` property. The default of:
```python
spec.name in spec.prefix.bin
```
is rather confusing, since the return type of that expression is a Boolean.

File: lib/spack/docs/packaging_guide.rst
Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst]: This got me stuck for quite a while.  I agree it should be implemented this way but there's something about the `libs` property that completely breaks down when you try to call another custom property inside it's implementation (no problem with the `headers` property).  The problem is unrelated to this change as I can reproduce it in current `develop` so I'll leave it as a TODO afterwards.

File: lib/spack/docs/packaging_guide.rst
Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst]: Re-worderd to:

"`spec.name` found in `spec.prefix.bin`"

File: var/spack/repos/builtin.mock/packages/attributes-foo/package.py
Comment (chuckatkins) [reply var/spack/repos/builtin.mock/packages/attributes-foo/package.py]: The layout is to specifically cover the use case of the `baz` virtual package being installed in a subdirectory of the prefix instead of at the top level of the prefix.

File: lib/spack/spack/test/install.py
Comment (chuckatkins) [reply lib/spack/spack/test/install.py:56]: If I simplify the test code some it's only 35 lines (I can drop the exception handling since that would appropriately generate a failure, the `libname` helper can be 2 lines, the comments are largely self evident and most can be dropped, etc.)

File: lib/spack/docs/packaging_guide.rst
Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst]: I'll drop the language around the new attribute being preferential.  For most use cases it only matters when directly querying a virtual dependency and even then only in the packages it actually gets used for. That can certainly be confusing when people often look to existing packages to see what they should do.

File: lib/spack/docs/packaging_guide.rst
Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst:2858]: There's two use cases in particular but they can't really be implemented until the feature is available so having the docs show an actual package implementing it is a chicken/egg problem.  The two use cases that motivated the pr (although there can certainly be others) are:

1. A package providing multiple components in sub-directories of it's install prefix making each component available as a virtual package.
2. Using dummy packages for complex scenarios that virtuals don't allow for (think virtuals providing virtuals).

1 is the easiest to understand and why I used it for the example.  It would be particularly useful for simplifying the intel and oneapi packages but is also useful for upcoming changes to paraview, vtk, and vtkm packages (vtk provides vtkm, and paraview provides vtk and vtkm).

2 is the more immediate use case and is being used to solve a structural issue with mesa, glx, and osmesa.  Mesa provides both glx and osmesa in the same build, and both of those have valid implementations of opengl.  Currently though there's no way to have mesa respond appropriately for `spec['gl'].libs` since there's no way of knowing *which* of the two `gl` implementations is being queried.  So I've created dummy `osmesa` and `glx` packages to allow the following:
```python
class Mesa(MesonPackage):
   provides('libosmesa', when='+osmesa')
   provides('libglx', when='+glx')


class Osmesa(BundlePackage):
    depends_on('libosmesa')
    provides('gl')

    @property
    def home(self):
        return self.spec['libosmesa'].prefix

    @property
    def headers(self):
        return self.spec['libosmesa'].headers

    @property
    def libs(self):
        return self.spec['libosmesa'].libs

    @property
    def gl_headers(self):
        return find_headers('GL/gl.h', root=self.home.include, recursive=False)


class Glx(BundlePackage):
    depends_on('libglx')
    provides('gl')

    @property
    def home(self):
        return self.spec['libglx'].prefix

    @property
    def headers(self):
        return self.spec['libglx'].headers

    @property
    def libs(self):
        return self.spec['libglx'].libs

    @property
    def gl_headers(self):
        return find_headers('GL/gl.h', root=self.home.include, recursive=False)
```

This effectively lets `mesa` provide two different `gl` virtual packages in the same build, each providing different `libs`

Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst:2858]: Regarding the second case, `spec['gl'].headers` and `.libs` work as is but `.prefix` will give the install path of the dummy pacakge which is not particularly helpful.  `.home` will give the correct install prefix delegated to `mesa` even though `gl` is provided by `osmesa` or `glx`.

File: lib/spack/docs/packaging_guide.rst
Comment (chuckatkins) [reply lib/spack/docs/packaging_guide.rst]: Okay, i was able to make this work by putting a default `home` property in the base package class similar to `prefix`

**Comment by tgamblin** [comment]
I may have missed this in the earlier discussion, but why do we need `home` *and* `prefix`?  Why can't `home` just be `prefix`?  I *believe* we can override that so that `spec['blas'].prefix` gives you what is being called `home` in this PR.  The question would be whether you ever want `spec[virtual].prefix` to refer to the prefix of the provider?  I could see it being confusing in core code to have this behavior, so maybe a separate name is better.

I think the fact that `home`, `prefix`, and `root` are usually synonyms isn't great for making this feature understandable.  If we need a special attribute, could the special attribute have a more descriptive name?  What about `.subprefix`?  I think that more clearly illustrates the concept than `home`.

**Comment by chuckatkins** [comment]
> I may have missed this in the earlier discussion, but why do we need `home` _and_ `prefix`? Why can't `home` just be `prefix`? I _believe_ we can override that so that `spec['blas'].prefix` gives you what is being called `home` in this PR. The question would be whether you ever want `spec[virtual].prefix` to refer to the prefix of the provider? I could see it being confusing in core code to have this behavior, so maybe a separate name is better.
> 

I started with overriding `.prefix` and got it to work but it was a bit weird and suffered from two problems specifically:

1. `prefix` already has a specific meaning across the core code and overriding that could be problematic in trying to access any metadata in the prefix's `.spack` folder.
2. `.prefix` takes different meanings for a "plain" `Spec` object as the on-disk install location vs as a derived `SpecBuildInterface` object with a customized implementation.

Both of the above seemed poised to create a stream of difficult to debug unforeseen consequences.

> I think the fact that `home`, `prefix`, and `root` are usually synonyms isn't great for making this feature understandable. If we need a special attribute, could the special attribute have a more descriptive name? What about `.subprefix`? I think that more clearly illustrates the concept than `home`.

I circled around on this for a good while trying to come up with a name.   Ultimately I ended up at `home` because it was already in use in a few packages for this very purpose (`python` and the various `java` providers).  `root` specifically is used in context of the DAG as an existing attribute which is why I changed it to `home`.  `subprefix` would be good if that was the only use case for it but there's several; a "subprefix" is just the most easiest to understand and what I wrote the test for.  Another is derived dummy packages as a way to create more featureful virtuals where `.home` is delegated to one of it's dependencies entirely outside of `prefix` (see https://github.com/spack/spack/pull/30917#discussion_r890389315), just to name one.

**Comment by tgamblin** [comment]
Ok given all that I think `prefix` probably isn't the way to go.  Do others thing that `home` is confusing or should we just go with it?  I think I still prefer `subprefix` or something a that's not usually completely synonymous with `prefix`, but if other folks are cool with this (@adamjstewart?) then I'm not going to put up a huge fight.

**Comment by adamjstewart** [comment]
I chose home long before this PR, mostly because of PYTHONHOME and JAVA_HOME, so I'm biased.

**Comment by chuckatkins** [comment]
Fixed test failures with the old concretizer

**Review by tgamblin — DISMISSED** [review]

**Comment by tgamblin** [comment]
ok looks like it's up to @alalazo now.

**Comment by chuckatkins** [comment]
Fixed style

**Comment by chuckatkins** [comment]
This should *actually* fix the original concretizer tests now.  Just simplified the test package to remove variants as it doesn't matter for what's actually being tested, it only needs to provide the virtuals.

**Comment by chuckatkins** [comment]
Updated `java` and `python` to use the new attribute to pass tests

**Comment by chuckatkins** [comment]
Ping @alalazo

**Review by alalazo — APPROVED** [review]
I still think the docs should be based on a real example, and test should be slightly simplified. Merging since this is a blocker for further work and core modifications LGTM (assuming `home` is fine with everybody as a name).

@chuckatkins Can you look into simplifying tests and docs in a later PR?

**Comment by chuckatkins** [comment]
> I still think the docs should be based on a real example, and test should be slightly simplified. Merging since this is a blocker for further work and core modifications LGTM (assuming `home` is fine with everybody as a name).
> 
> @chuckatkins Can you look into simplifying tests and docs in a later PR?

Once merged I'll apply this to the GL stack in the OpenGL, Mesa, OSMesa, and GLX packages and can then use those specifically in the docs to demonstrate the use cases.

**Pull request merged**
